core/lib/zksync_core/src/basic_witness_input_producer/mod.rs

```diff
@@ -0,0 +1,307 @@
+        prev_block_hash: miniblock_hash,
+        max_virtual_blocks_to_create: virtual_blocks,
+    }
+}
+
+fn execute_tx<S: ReadStorage>(tx: &Transaction, vm: &mut VmInstance<S, HistoryEnabled>) {
+    vm.make_snapshot();
+    if vm
+        .inspect_transaction_with_bytecode_compression(vec![], tx.clone(), true)
+        .is_ok()
+    {
+        vm.pop_snapshot_no_rollback();
+        return;
+    }
+    vm.rollback_to_the_latest_snapshot();
+    vm.inspect_transaction_with_bytecode_compression(vec![], tx.clone(), false)
+        .expect("Compression can't fail if we don't apply it");
+}
+
+pub struct BasicWitnessInputProducer {
+    connection_pool: ConnectionPool,
+    validation_computational_gas_limit: u32,
+    l2_chain_id: L2ChainId,
+    object_store: Arc<dyn ObjectStore>,
+}
+
+impl BasicWitnessInputProducer {
+    pub async fn new(
+        connection_pool: ConnectionPool,
+        store_factory: &ObjectStoreFactory,
+        state_keeper_config: &StateKeeperConfig,
+        network_config: &NetworkConfig,
+    ) -> anyhow::Result<Self> {
+        // BEWARE, HERE BE DRAGONS.
+        // This usage here is a race condition waiting to happen.
+        // Context -- The VM run in BasicWitnessInputProducer MUST be identical to run from StateKeeper.
+        // Race condition scenario:
+        // 1. StateKeeper runs with a config A, which has value a for `validation_computational_gas_limit`
+        // 2. StateKeeper runs the batch and saves data to database.
+        // 3. Deployment is made and changes to config B, with value b for `validation_computational_gas_limit`.
+        // 4. BasicWitnessInputProducer loads config B
+        // 5. BasicWitnessInputProducer runs the L1Batch that was saved with config A using config B.
+        // In this scenario, the outputs will be different, which will make Witness Generation crash.
+        // This has been discussed between @evl and @deniallugo as a terrible solution, but best so far.
+        let validation_computational_gas_limit =
+            state_keeper_config.validation_computational_gas_limit;
+        // This variable is fine to be loaded from config.
+        // Today, this config changes only during regenesis.
+        // In the hyper-chains world, this code will be refactored (using the DB saved value instead of config)
+        let l2_chain_id = network_config.zksync_network_id;
+
+        Ok(BasicWitnessInputProducer {
+            connection_pool,
+            validation_computational_gas_limit,
+            object_store: store_factory.create_store().await.into(),
+            l2_chain_id,
+        })
+    }
+
+    async fn process_job_impl(
+        job: BasicWitnessInputProducerJob,
+        started_at: Instant,
+        connection_pool: ConnectionPool,
+        validation_computational_gas_limit: u32,
+        l2_chain_id: L2ChainId,
+    ) -> anyhow::Result<WitnessBlockState> {
+        let l1_batch_number = job.l1_batch_number;
```

Comment:
nit: I personally don't see much value in this local variable. `job.l1_batch_number` is perfectly readable, and arguably carries more semantical meaning (you **know** what kind of l1 batch number the author refers to - the one from the `job`. I would remove this line